pkg/backup/backup_new_test.go

```diff
@@ -0,0 +1,851 @@
+
+	h.backupper.Backup(h.log, backup2, backup2File, nil, nil)
+
+	assertTarballContents(t, backup2File, "metadata/version", "resources/deployments.apps/namespaces/ns-1/deploy-1.json")
+}
+
+func TestBackupResourceOrdering(t *testing.T) {
+	tests := []struct {
+		name         string
+		backup       *velerov1.Backup
+		apiResources []*apiResource
+	}{
+		{
+			name: "core API group: pods come before pvcs, pvcs come before pvs, pvs come before anything else",
+			backup: defaultBackup().
+				SnapshotVolumes(false).
+				Build(),
+			apiResources: []*apiResource{
+				pods(
+					newPod("foo", "bar"),
+					newPod("zoo", "raz"),
+				),
+				pvcs(
+					newPVC("foo", "bar"),
+					newPVC("zoo", "raz"),
+				),
+				pvs(
+					newPV("bar"),
+					newPV("baz"),
+				),
+				secrets(
+					newSecret("foo", "bar"),
+					newSecret("zoo", "raz"),
+				),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				h          = newHarness(t)
+				req        = &Request{Backup: tc.backup}
+				backupFile = bytes.NewBuffer([]byte{})
+			)
+
+			for _, resource := range tc.apiResources {
+				h.addItems(t, resource.group, resource.version, resource.name, resource.shortName, resource.namespaced, resource.items...)
+			}
+
+			h.backupper.Backup(h.log, req, backupFile, nil, nil)
+
+			assertTarballOrdering(t, backupFile, "pods", "persistentvolumeclaims", "persistentvolumes")
+		})
+	}
+}
+
+type apiResource struct {
+	group      string
+	version    string
+	name       string
+	shortName  string
+	namespaced bool
+	items      []metav1.Object
+}
+
+func pods(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "",
+		version:    "v1",
+		name:       "pods",
+		shortName:  "po",
+		namespaced: true,
+		items:      items,
+	}
+}
+
+func pvcs(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "",
+		version:    "v1",
+		name:       "persistentvolumeclaims",
+		shortName:  "pvc",
+		namespaced: true,
+		items:      items,
+	}
+}
+
+func secrets(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "",
+		version:    "v1",
+		name:       "secrets",
+		shortName:  "secrets",
+		namespaced: true,
+		items:      items,
+	}
+}
+
+func deployments(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "apps",
+		version:    "v1",
+		name:       "deployments",
+		shortName:  "deploy",
+		namespaced: true,
+		items:      items,
+	}
+}
+
+func extensionsDeployments(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "extensions",
+		version:    "v1",
+		name:       "deployments",
+		shortName:  "deploy",
+		namespaced: true,
+		items:      items,
+	}
+}
+
+func pvs(items ...metav1.Object) *apiResource {
+	return &apiResource{
+		group:      "",
+		version:    "v1",
+		name:       "persistentvolumes",
+		shortName:  "pv",
+		namespaced: false,
+		items:      items,
+	}
+}
+
+type harness struct {
```

Comment:
I think some of it may move out, yeah. I need to feel it out a little as I move on to working on `pkg/restore`.  Either way, anything that does become shared, I'm going to move into `pkg/test`, with the goal of eventually getting rid of `pkg/util/test` (anything still needed will migrate).

Reply:
Cool. No reason to move it now as we don't know where it's actually shared, just checking.